File: core/LinkedListCycleII/LinkedListCycleII.java
package com.dbc;

import java.util.HashMap;
import java.util.Map;

public class LinkedListCycleII {
    class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public ListNode detectCycle(ListNode head) {
        Map<ListNode, Boolean> map = new HashMap<>();
        while (head != null) {
            if (map.getOrDefault(head, false)) return head;
            map.put(head, true);
            head = head.next;
        }
        return null;
    }
}

File: core/LinkedListCycleII/LinkedListCycleII.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def detectCycle(head: ListNode) -> ListNode:
    remain = {}
    while head:
        if remain.get(head, False):
            return head
        remain[head] = True
        head = head.next

    return None

def detectCycle(head: ListNode) -> ListNode:
    slow, fast = head, head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            break

    if not fast or not fast.next:
        return None
    pre_head = head
    while pre_head != slow:
        pre_head = pre_head.next
        slow = slow.next

    return pre_head




